### PR TITLE
vimc-2174

### PR DESCRIPTION
**Pull request by JamesThompson1729**

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -35,6 +35,7 @@ Imports:
     yaml,
     zip (>= 2.0.0)
 Suggests:
+    crayon,
```
Comment (richfitz) [inline DESCRIPTION]: This is being used unconditionally so needs to be in Imports

File: R/dependencies.R [hunk 2]
```diff
@@ -0,0 +1,410 @@
+##' @title Get the dependencies for a given report from the database
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param con A connection to a database
+get_dependencies_db <- function(name, id, upstream, con, list_all = FALSE) {
```
Comment (richfitz) [inline R/dependencies.R]: All exported orderly functions have `orderly_` as their name, and in general I do not like the use of `get` here. How about simply `orderly_dependencies` (or perhaps `orderly_report_dependencies`)

File: R/dependencies.R [hunk 3]
```diff
@@ -0,0 +1,410 @@
+##' @title Get the dependencies for a given report from the database
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param upstream A boolean indicating if we want to move up or down the tree
```
Comment (richfitz) [inline R/dependencies.R]: is there a case where we might want to do *both* directions? In that case, even if not implemented now we'd be better off with a character argument here (`direction = "upstream"` or `direction = "downstream"`)

File: R/dependencies.R [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+##' @param con A connection to a database
+get_dependencies_db <- function(name, id, upstream, con, list_all = FALSE) {
+  ## now construct the SQL query
+  if (upstream) {
+    filt_qry <- sprintf("depends.report_version='%s'", id)
```
Comment (richfitz) [inline R/dependencies.R]: please avoid excessive abbreviation (filt, qry, etc)

File: R/dependencies.R [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+  } else {
+    filt_qry <- sprintf("report_version.id='%s'", id)
+  }
+
+  sql_qry <- paste(c("SELECT",
```
Comment (richfitz) [inline R/dependencies.R]: Can we move this into the sql directory in inst? If not, there seem to be more commas here than needed

File: R/dependencies.R [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+  db_ret <- DBI::dbGetQuery(con, sql_qry)
+
+  if (nrow(db_ret) == 0) {
+    return(NULL)
+  }
+
+  db_ret$is_latest <- sapply(db_ret$report_version,
+                               is_latest_in_db, con = con)
+  if (!list_all) {
+    db_ret <- db_ret[which(db_ret$is_latest), ]
```
Comment (richfitz) [inline R/dependencies.R]: no need for `which` here

File: R/dependencies.R [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+               "report_version_artefact.report_version=report_version.id)",
+               "WHERE", filt_qry), collapse = " ")
+
+  db_ret <- DBI::dbGetQuery(con, sql_qry)
+
+  if (nrow(db_ret) == 0) {
+    return(NULL)
+  }
+
+  db_ret$is_latest <- sapply(db_ret$report_version,
```
Comment (richfitz) [inline R/dependencies.R]: please avoid sapply - here use `vlapply`

File: R/dependencies.R [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+                               is_latest_in_db, con = con)
+  if (!list_all) {
+    db_ret <- db_ret[which(db_ret$is_latest), ]
+  }
+
+  # if we're going uptree
+  if (upstream) {
+    return(unique(db_ret$id))
```
Comment (richfitz) [inline R/dependencies.R]: please do not use explicit return statements if it is not an early exit (here and elsewhere)

File: R/dependencies.R (hunk 2 above)
Comment (richfitz) [reply R/dependencies.R]: oh, I see this is not actually an exported function - in which case I think these roxygen blocks all require `@noRd` or similar to prevent creation of `man` files

File: R/dependencies.R [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+  } else {
+    return(unique(db_ret$report_version))
+  }
+}
+
+##' @title Get the id of the latest version of a report
+##'
+##' @param name the name of the report
+##' @param con A connection to a database
+get_latest_by_name <- function(con, name) {
```
Comment (richfitz) [inline R/dependencies.R]: The `get` here is not helpful - something like `report_latest_version` might capture intent better.  However we have similar functions in query.R already - could you harmonise?

File: R/dependencies.R [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+  sql_qry <- paste("SELECT", "id, report, date FROM report_version",
+                   "WHERE", sprintf("report='%s'", name),
+                   "AND", "date=(SELECT MAX(date)",
+                   "FROM", "report_version",
+                   sprintf("WHERE report='%s')", name))
+  db_ret <- DBI::dbGetQuery(con, sql_qry)
+  return(db_ret)
+}
+
+get_latest_by_id <- function(con, id) {
```
Comment (richfitz) [inline R/dependencies.R]: this could be harmonised with code in (query.R & nearby functions)

File: R/dependencies.R [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+  }
+}
+
+##' @title Is the id the latest version of the report in the database
+##'
+##' @param id the id of the report
+##' @param con A connection to a database
+##'
+##' @return A boolean TRUE if the report is the latest version
+is_latest_in_db <- function(con, id) {
```
Comment (richfitz) [inline R/dependencies.R]: do these all warrant being free functions?

File: R/dependencies.R [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+
+  db_chd <- DBI::dbGetQuery(con, sql_chd_qry)
+
+  ## this reports uses no artefacts - so can never be out of date
+  if (nrow(db_chd) == 0) {
+    return(FALSE)
+  }
+
+  ## we iterate over rows of the data frame and make sure the artefact match
+  for (i in 1:nrow(db_chd)) {
```
Comment (richfitz) [inline R/dependencies.R]: please do not use `1:n` for sequences - use `seq_len`

File: R/dependencies.R [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+    filename <- db_chd$filename[i]
+    file_hash <- db_chd$file_hash[i]
+    report_id <- db_chd$report_version[i]
+    ## we need to find the id latest version of the report with id = report_id
+    if (!db_chd$is_pinned[i]) { ## if not pinned...
+      latest_id <- get_latest_by_id(con, report_id) ## ..use latest
+    } else { ## if pinned to..
+      if (!db_chd$is_latest[i]) { ## ..not latest..
```
Comment (richfitz) [inline R/dependencies.R]: this if/else has the same body in both branches

File: R/dependencies.R [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+             "FROM", "(report_version",
+             "INNER JOIN", "report_version_artefact", "ON",
+                           "report_version.id = report_version_artefact.report_version",
+             "INNER JOIN", "file_artefact", "ON",
+                           "file_artefact.artefact = report_version_artefact.id)",
+             "WHERE", par_qry), collapse = " ")
+
+    db_par <- DBI::dbGetQuery(con, sql_par_qry)
+    i <- which(db_par$filename == filename)
+    if (db_par$file_hash[i] != file_hash)
```
Comment (richfitz) [inline R/dependencies.R]: please always use braces, even for one-line statements. The `which` in the line above appears unneeded

File: R/dependencies.R [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+
+##' @title Recursively check that none of the parents share the same name as the
+##' current report.
+##'
+##' @param parent_vertex The vertex whose name we want to check
+##' @param name The name we want to match
+##'
+##' @return A boolean TRUE if there is a match (bad!) FALSE if the is no match
+##' (good!)
+check_parents <- function(parent_vertex, name) {
```
Comment (richfitz) [inline R/dependencies.R]: Can you elaborate on when this is `FALSE`?

File: R/dependencies.R [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+##' @param locate
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##'
+##' @return An R6 tree object
+build_tree <- function(name, id, depth = 100, parent = NULL,
+                       tree = NULL, con, upstream = FALSE, list_all = FALSE) {
+  ## this should never get triggered - it only exists the prevent an infinite
+  ## recursion
+  if (depth < 0) {
+    stop("The tree is very large or degenerate.")
```
Comment (richfitz) [inline R/dependencies.R]: This is not a helpful message for users - what leads to this situation? Why is large a problem?

File: R/dependencies.R [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+  if (!is.null(parent)) {
+    circ <- check_parents(parent, name)
+    if (circ) {
+      tree$message <- "There appears to be a circular dependency."
```
Comment (richfitz) [inline R/dependencies.R]: is this an error condition?

File: R/dependencies.R [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+      return(tree)
+    }
+  }
+
+  # make sure a report with this name exists
+  db_reports <- DBI::dbGetQuery(con, "SELECT name FROM report")
+  if (!(name %in% db_reports$name)) {
+    stop("This report does not exist")
```
Comment (richfitz) [inline R/dependencies.R]: Can this happen? If we're working purely with the DB then the foreign keys are doing all the referential integrity work for us.

File: R/dependencies.R [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+  # do we need to find the latest version of the report?
+  if (id == "latest") {
+    id <- get_latest_by_name(con, name)$id
```
Comment (richfitz) [inline R/dependencies.R]: everywhere you use `get_latest_by_name` you are extracting `$id` so probably refactor the function to just return that

File: R/dependencies.R [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+  } else {
+    db_id <- id_to_name(con, id)
+    if (is.null(db_id)) {
+      stop(sprintf("No report with id %s in the database", id))
+    } else {
```
Comment (richfitz) [inline R/dependencies.R]: collapse the `else` and the `if`

File: R/dependencies.R [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+      if (name != id_to_name(con, id)) {
+        stop(sprintf("id %s does not match report name %s", id, name))
+      }
+    }
+  }
+
+  if (upstream) {
```
Comment (richfitz) [inline R/dependencies.R]: both branches have the same content?

File: R/dependencies.R [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+      str <- paste(str, crayon::red(self$message), "\n", collapse = "")
+      self$message <- NULL
+    }
+
+    console_colour <- if (vertex$out_of_date) {crayon::red} else {crayon::blue}
+
+    if (length(fvector) == 0) {
+      str <- paste(str, console_colour(sprintf("%s", vertex$format())), "\n", collapse = "")
+    } else {
+      spacing <- paste(ifelse(head(fvector, -1), "| ", "  "), collapse = "")
```
Comment (richfitz) [inline R/dependencies.R]: do not use ifelse

File: R/dependencies.R [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+  },
+  # this is stupidly hacky to get the formatting right
+  format_helper = function(vertex = self$root, fvector = c(), str="") {
+    ## if the tree has a warning message, append it to the front in the red
+    if (!is.null(self$message)) {
+      str <- paste(str, crayon::red(self$message), "\n", collapse = "")
+      self$message <- NULL
+    }
+
+    console_colour <- if (vertex$out_of_date) {crayon::red} else {crayon::blue}
```
Comment (richfitz) [inline R/dependencies.R]: no braces on a one-line if/else

File: R/dependencies.R [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+Tree <- R6::R6Class("Tree", list(
+  vertices = NULL,
+  root = NULL,
+  message = NULL,
+  initialize = function(root) {
+    self$vertices <- append(self$vertices, list(root))
+    self$root <- root
+    self$message <- NULL
+  },
+  add_child = function(parent, name, id, out_of_date) {
```
Comment (richfitz) [inline R/dependencies.R]: one blank line between all methods

File: R/dependencies.R [hunk 25]
```diff
@@ -0,0 +1,410 @@
...
+      if (vertex$out_of_date) {
+        child$out_of_date = TRUE
+      }
+    }
+
+    propagate(child, upstream)
+
+    if (upstream) {
+      if (child$out_of_date) {
+        vertex$out_of_date = TRUE
```
Comment (richfitz) [inline R/dependencies.R]: no equals for assignment

File: R/dependencies.R [hunk 26]
```diff
@@ -0,0 +1,410 @@
...
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param con A connection to a database
+##' @param root
+##' @param locate
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param propagate A boolean indicating if we want to propogate out of date
+##'                  through the tree
+##'
+##' @export
```
Comment (richfitz) [inline R/dependencies.R]: This has not yet been exported - you need to re-document the package

File: R/dependencies.R [hunk 27]
```diff
@@ -0,0 +1,410 @@
...
+orderly_build_dep_tree <- function(name, id = "latest", root = NULL,
```
Comment (richfitz) [inline R/dependencies.R]: `orderly_dependency_tree` perhaps - avoids the abbreviation and the `build` part is not needed

File: R/dependencies.R [hunk 28]
```diff
@@ -0,0 +1,410 @@
...
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param con A connection to a database
+##' @param root
+##' @param locate
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param propagate A boolean indicating if we want to propogate out of date
+##'                  through the tree
+##'
```
Comment (richfitz) [inline R/dependencies.R]: please add an example - you can use the demo orderly example

File: R/dependencies.R [hunk 29]
```diff
@@ -0,0 +1,410 @@
...
+    for (i in 1:length(vertex$children)) {
+      child <- vertex$children[[i]]
+
+      reports <- out_ot_date_reports(child, reports)
+    }
+  }
+  reports
+}
+
+##' @title Print the dependency tree for a given report using orderly log
```
Comment (richfitz) [inline R/dependencies.R]: Please move this to the top of the file

File: R/dependencies.R [hunk 30]
```diff
@@ -0,0 +1,410 @@
...
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param con A connection to a database
+##' @param root
```
Comment (richfitz) [inline R/dependencies.R]: Use @inheritParams to get these added - see `orderly_run` for example

File: R/dependencies.R [hunk 31]
```diff
@@ -0,0 +1,410 @@
...
+      child <- vertex$children[[i]]
+
+      reports <- out_ot_date_reports(child, reports)
+    }
+  }
+  reports
+}
+
+##' @title Print the dependency tree for a given report using orderly log
+##'
```
Comment (richfitz) [inline R/dependencies.R]: Please expand the discussion of what this function is for, how one uses it, etc

File: R/dependencies.R [hunk 32]
```diff
@@ -0,0 +1,410 @@
...
+    } else {
+      spacing <- paste(ifelse(head(fvector, -1), "| ", "  "), collapse = "")
+      str <- paste(str, console_colour(sprintf("%s|___%s", spacing, vertex$format())), "\n", collapse = "")
+    }
+
+    if (length(vertex$children) > 0) {
+      for (i in 1:length(vertex$children)) {
+        child <- vertex$children[[i]]
+
+        if (i != length(vertex$children)) {
```
Comment (richfitz) [inline R/dependencies.R]: this mgiht be simpler as

```
thing <- i != length(vertex$children)
str <- self$format_helper(child, c(fvector, thing), str)
```

though I am unclear what the second argument here represents at this point

File: R/dependencies.R [hunk 33]
```diff
@@ -0,0 +1,410 @@
...
+    self$message <- NULL
+  },
+  add_child = function(parent, name, id, out_of_date) {
+    child <- Vertex$new(parent, name, id, out_of_date)
+    parent$add_child(child)
+    self$vertices <- append(self$vertices, list(child))
+    return(child)
+  },
+  # this is stupidly hacky to get the formatting right
+  format_helper = function(vertex = self$root, fvector = c(), str="") {
```
Comment (richfitz) [inline R/dependencies.R]: this should be a private method, and we should think about making it not stupidly hacky, or at least explaining what the problem is

File: R/dependencies.R [hunk 34]
```diff
@@ -0,0 +1,410 @@
...
+
+    if (upstream) {
+      if (child$out_of_date) {
+        vertex$out_of_date = TRUE
+      }
+    }
+  }
+}
+
+Vertex <- R6::R6Class("Vertex", list(
```
Comment (richfitz) [inline R/dependencies.R]: this and tree seem totally unrelated to everything else in the file, which is good. Can you move them to their own file and add some testing to them separate to the rest of the dependency code

File: R/dependencies.R [hunk 35]
```diff
@@ -0,0 +1,410 @@
...
+  for (child in vertex$children) {
+    if (!upstream) {
+      if (vertex$out_of_date) {
+        child$out_of_date = TRUE
+      }
+    }
+
+    propagate(child, upstream)
+
+    if (upstream) {
```
Comment (richfitz) [inline R/dependencies.R]:
```
if (upstream && child$out_of_date) {
  vertex$out_of_date <- TRUE
}
```

or even

```
vertex$out_of_date <- upstream && child$out_of_date
```

and similar above. there's quite a bit of noise that can be removed like this

File: R/dependencies.R [hunk 36]
```diff
@@ -0,0 +1,410 @@
...
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param propagate A boolean indicating if we want to propogate out of date
+##'                  through the tree
+##'
+##' @export
+orderly_build_dep_tree <- function(name, id = "latest", root = NULL,
+                                   locate = TRUE, upstream = FALSE, con = NULL,
+                                   propagate = FALSE, max_depth = 100,
+                                   list_all = FALSE) {
+  if (is.null(con)) {
```
Comment (richfitz) [inline R/dependencies.R]: I don't see a use case for accepting both root/locate and con - drop the con argument and make the connection unconditional

**Comment by JamesThompson1729** [comment]
I have made all the changes.
We use the the local report database to find dependencies, this causes a few problems 
 * If a report has been rewritten to change the dependencies this wont get picked up since we never read the yaml files.
 * If a report has been re-run nothing will depend on it. To get around this I added a _"previous"_ flag to _"id"_ which gets the id of the second most recent version of a report.

The two basic use cases are:
 * We want to update a report and need to know which _'upstream'_ reports it depends on - 
`print(orderly_build_dep_tree("report_name", direction = "upstream"))`
This will print the dependency tree with out-of-date reports marked in red.

 * We have re-run and report and we want to re-run everything that depends on it -
`print(orderly_build_dep_tree("report_name", direction = "downstream", id = "previous"))`
we have to use the _previous_ because once a report has been re-run nothing depends on yet. So we check what depended on the previous version of the report.

File: R/dependencies.R (hunk 31 above)
Comment (richfitz) [reply R/dependencies.R]: Still pending

File: R/dependencies.R (hunk 28 above)
Comment (richfitz) [reply R/dependencies.R]: Still pending

File: R/dependencies.R (hunk 26 above)
Comment (richfitz) [reply R/dependencies.R]: still pending

File: R/dependencies.R (hunk 27 above)
Comment (richfitz) [reply R/dependencies.R]: Still pending

File: R/dependencies.R (hunk 16 above)
Comment (JamesThompson1729) [reply R/dependencies.R]: This was to protect against a scenario where the tree somehow has circular dependency.
I could not guarentee that this impossible. and I wanted to avoid a situation where we get into an infinite recursion where orderly hangs and the user has no idea what went wrong.

File: R/dependencies.R (hunk 18 above)
Comment (JamesThompson1729) [reply R/dependencies.R]: I think it can only be hit if the user passes in an invalid name. I've moved this logic to *orderly_build_dep_tree*, so we only check once when before the recursive steps begin.

File: R/dependencies.R [hunk 37]
```diff
@@ -0,0 +1,471 @@
...
+##'    on it.
+##' 2. We have a report that we want to re-run to ensure uses the latest
+##'    information. Which other reports are used (directly or indirectly) by
+##'    this report?
+##'
+##' This function display this information in an easily readable format.
+##' Allowing users to see the dependency tree and which reports are out of date 
+##' and need to be re-run.
+##'
+##' Remark
```
Comment (richfitz) [inline R/dependencies.R]:
```
##' @section Remark:
```

File: R/dependencies.R [hunk 38]
```diff
@@ -0,0 +1,471 @@
...
+##' @title Given a tree return a list of reports to be re-run (and the order
+##' that they should be re-run)
+##'
+##' @param tree A dpendency tree object from orderly_dependency_tree
+##'
+##' @return a list of report names to be re-run. First report to rerun first
+##' @export
+orderly_out_of_date_reports <- function(tree) {
+  types <- class(tree)
+  if (!("Tree" %in% types)) {
```
Comment (richfitz) [inline R/dependencies.R]: `assert_is(tree, "Tree")`

File: R/dependencies.R [hunk 39]
```diff
@@ -0,0 +1,471 @@
...
+
+  sql_query <- paste(c(sql_query_file, "WHERE", filter_query), collapse = " ")
+
+  query_return <- DBI::dbGetQuery(con, sql_query)
+
+  if (nrow(query_return) == 0) {
+    return(NULL)
+  }
+
+  query_return$is_latest <- vapply(query_return$report_version,
```
Comment (richfitz) [inline R/dependencies.R]: vlapply

File: R/dependencies.R [hunk 40]
```diff
@@ -0,0 +1,471 @@
...
+##' @return a one line dataframe with columns:
+##'         id - the report id
+##'         report - the name of the report (matches the argument name)
+##'         date - the date the report was run
+##'
+##' @noRd
+get_latest_by_name <- function(con, name) {
+  reports <- get_ids_by_name(con, name)
+
+  (reports$id)[1]
```
Comment (richfitz) [inline R/dependencies.R]: no need for parens here

File: R/dependencies.R [hunk 41]
```diff
@@ -0,0 +1,471 @@
...
+##'
+##' @noRd
+get_previous_by_name <- function(con, name) {
+  reports <- get_ids_by_name(con, name)
+
+  if (nrow(reports) < 2) {
+    stop(sprintf("There is only one version of %s", name))
+  }
+
+  (reports$id)[2]
```
Comment (richfitz) [inline R/dependencies.R]: No parens
